plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java

```diff
@@ -175,50 +197,81 @@ private String determinePageLocation(IServer server, ILaunchConfiguration config
+   * <li>On server stop, terminate the launch</li>
+   * </ul>
    */
-  private class OpenBrowserListener implements IServerListener {
-    private String pageLocation;
+  private class ServerLaunchMonitor implements ILaunchesListener2, IServerListener {
+    private ILaunchConfiguration configuration;
+    private ILaunch launch;
+    private IServer server;
+
+    /**
+     * Setup the monitor.
+     */
+    ServerLaunchMonitor(ILaunchConfiguration configuration, ILaunch launch, IServer server) {
+      this.configuration = configuration;
+      this.launch = launch;
+      this.server = server;
+    }
 
-    public OpenBrowserListener(String pageLocation) {
-      this.pageLocation = pageLocation;
+    private void engage() {
+      getLaunchManager().addLaunchListener(this);
+      server.addServerListener(this);
+    }
+
+    private void disengage() {
+      getLaunchManager().removeLaunchListener(this);
+      server.removeServerListener(this);
     }
 
     @Override
     public void serverChanged(ServerEvent event) {
-      if (event.getState() != IServer.STATE_STARTED) {
-        return;
-      }
-      final IServer server = event.getServer();
-      final IWorkbench workbench = PlatformUI.getWorkbench();
+      assert server == event.getServer();
+      switch (event.getState()) {
+        case IServer.STATE_STARTED:
+          openBrowserPage(configuration, server);
+          return;
 
-      Job openJob = new UIJob(workbench.getDisplay(), "Launching start page") {
-
-        @Override
-        public IStatus runInUIThread(IProgressMonitor monitor) {
-          if (server.getServerState() != IServer.STATE_STARTED) {
-            return Status.CANCEL_STATUS;
-          }
+        case IServer.STATE_STOPPED:
+          disengage();
           try {
-            URL url = new URL(pageLocation);
-            IWorkbenchBrowserSupport browserSupport = workbench.getBrowserSupport();
-            int style = IWorkbenchBrowserSupport.LOCATION_BAR
-                | IWorkbenchBrowserSupport.NAVIGATION_BAR | IWorkbenchBrowserSupport.STATUS;
-            browserSupport.createBrowser(style, server.getId(), server.getName(), server.getName())
-                .openURL(url);
-          } catch (PartInitException ex) {
-            // Unable to use the normal browser support, so punt to the OS
-            logger.log(Level.WARNING, "Cannot launch a browser", ex);
-            Program.launch(pageLocation);
-          } catch (MalformedURLException ex) {
-            logger.log(Level.WARNING, "Unable to determine dev_appserver URL", ex);
+            logger.fine("Server stopped; terminating launch");//$NON-NLS-1$
+            launch.terminate();
+          } catch (DebugException ex) {
+            logger.log(Level.WARNING, "Unable to terminate launch", ex);
           }
-          return Status.OK_STATUS;
+      }
+    }
+
+    @Override
+    public void launchesTerminated(ILaunch[] launches) {
+      for (ILaunch l : launches) {
```

Comment:
please don't use l as local variable name; depending on font it may not be readable; maybe launch and this.launch?